src/services/createRouter.ts

```diff
@@ -182,72 +187,79 @@ export function createRouter<const TRoutes extends Routes, const TOptions extend
   }
 
   const resolve: RouterResolve<TRoutes> = (
-    source: Url | RoutesName<TRoutes>,
-    paramsOrOptions?: Record<string, unknown>,
-    maybeOptions?: RouterResolveOptions,
+    source: RoutesName<TRoutes>,
+    params: Record<string, unknown> = {},
+    options: RouterResolveOptions = {},
   ) => {
-    if (!isUrl(source)) {
-      const params = paramsOrOptions ?? {}
-      const options: RouterResolveOptions = maybeOptions ?? {}
-      const match = routes.find((route) => route.name === source)
+    const match = routes.find((route) => route.name === source)
 
-      if (!match) {
-        return undefined
-      }
-
-      return createResolvedRoute(match, params, options)
-    }
-
-    if (isExternal(source)) {
-      return undefined
+    if (!match) {
+      throw new RouteNotFoundError(String(source))
     }
 
-    const options: RouterPushOptions = { ...paramsOrOptions }
-    const url = appendQuery(source, options.query)
-
-    return createResolvedRouteForUrl(routes, url)
+    return createResolvedRoute(match, params, options)
   }
 
-  const push: RouterPush<TRoutes> = (source: Url | RoutesName<TRoutes>, paramsOrOptions?: Record<string, unknown> | RouterPushOptions, maybeOptions?: RouterPushOptions) => {
+  const push: RouterPush<TRoutes> = (
+    source: Url | RoutesName<TRoutes> | ResolvedRoute,
+    paramsOrOptions?: Record<string, unknown> | RouterPushOptions,
+    maybeOptions?: RouterPushOptions,
+  ) => {
     if (isUrl(source)) {
       const options: RouterPushOptions = { ...paramsOrOptions }
-      const url = appendQuery(source, options.query)
+      const resolved = find(source, options)
+      const mightBeExternal = resolved === undefined
+
+      if (mightBeExternal) {
+        const url = combineUrl(source, {
+          searchParams: new URLSearchParams(options.query),
+          hash: options.hash,
+        })
+
+        return set(url, options)
+      }
 
-      return set(url, options)
+      const { replace, state } = options
+
+      return push(resolved, { replace, state })
     }
 
-    const options: RouterPushOptions = { ...maybeOptions }
-    const params: any = { ...paramsOrOptions }
-    const resolved = resolve(source, params, options)
+    if (typeof source === 'string') {
+      const options: RouterPushOptions = { ...maybeOptions }
+      const params: any = { ...paramsOrOptions }
+      const resolved = resolve(source, params, options)
+      const { replace, state } = options
 
-    if (!resolved) {
-      throw new RouteNotFoundError(String(source))
+      return push(resolved, { replace, state })
     }
 
-    const state = setStateValues({ ...resolved.state }, options.state)
+    const { replace, ...options }: RouterPushOptions = { ...paramsOrOptions }
+    const state = setStateValues({ ...source.matched.state }, { ...source.state, ...options.state })
```

Comment:
My guess is this is written recursively to avoid repeating this setStateValues logic. 

I think it would actually be simpler to not be recursive and just handle each of the three possibilities individually rather than attempting to normalize to a resolved route.